File: full_case_study_solns/_d_mass/simulink/hw11/massParamHW11.m
% tuning parameters
tr = 2;         % rise time from D.8
% tr = 2.5;       % slower rise time to avoid saturation
zeta = 0.7;

% sample time for controller
P.Ts = 0.01;

% dirty derivative gain
P.sigma = 0.05;
P.beta = (2*P.sigma-P.Ts)/(2*P.sigma+P.Ts); % dirty derivative gain

% input constraint
P.Fmax = 2;

% state space design 
A = [...
    0, 1;...
    -P.k/P.m, -P.b/P.m;...
    ];
B = [0; 1/P.m ];
C = [...
    1, 0;...
    ];

% gain calculation
wn = 2.2/tr;
ol_char_poly = charpoly(A);
des_char_poly = [1,2*zeta*wn,wn^2];
des_poles = roots(des_char_poly);
% is the system controllable?
if rank(ctrb(A,B))~=2, disp('System Not Controllable'); end
P.K = place(A,B,des_poles);
P.kr = -1/(C*inv(A-B*P.K)*B);

File: full_case_study_solns/_d_mass/python_old/hw11/massParamHW11.py
# Single link mass Parameter File
import numpy as np
import control as cnt
import sys
sys.path.append('..')  # add parent directory
import massParam as P

Ts = P.Ts  # sample rate of the controller
beta = P.beta  # dirty derivative gain
F_max = P.F_max  # limit on control signal

#  tuning parameters
tr = 2.0
zeta = 0.707

# State Space Equations
# xdot = A*x + B*u
# y = C*x
A = np.matrix([[0.0, 1.0],
               [-P.k/P.m, -P.b/P.m]])

B = np.matrix([[0.0],
               [1.0/P.m]])

C = np.matrix([[1.0, 0.0]])

# gain calculation
wn = 2.2/tr  # natural frequency
des_char_poly = [1, 2*zeta*wn, wn**2]
des_poles = np.roots(des_char_poly)

# Compute the gains if the system is controllable
if np.linalg.matrix_rank(cnt.ctrb(A, B)) != 2:
    print("The system is not controllable")
else:
    K = cnt.acker(A, B, des_poles)
    kr = -1.0/(C[0]*np.linalg.inv(A-B*K)*B)

print('K: ', K)
print('kr: ', kr)



